Fix pkcs8 unencrypted private key parsing with Attributes field

BhanuPrakash-P opened the pull request:
## Description

Fix pkcs8 unencrypted private key parsing failure due to attributes field present in the EC private key.
Fixes https://github.com/Mbed-TLS/mbedtls/issues/9422

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [ ] **changelog** provided | not required because: 
- [ ] **development PR** provided # | not required because: 
- [ ] **framework PR** provided Mbed-TLS/mbedtls-framework# | not required
- [ ] **3.6 PR** provided # | not required because: 
- [ ] **2.28 PR** provided # | not required because: 
- **tests**  provided | not required because:

gilles-peskine-arm commented:
The CI is complaning of misplaced whitespace. Can you please fix those before we review the change?

Also, since this is a bug fix, please provide a non-regression test case.

BhanuPrakash-P commented:
> The CI is complaning of misplaced whitespace. Can you please fix those before we review the change?
> 
> Also, since this is a bug fix, please provide a non-regression test case and add a [changelog entry](https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/00README.md).

I have addressed CI failures and updated the ChangeLog.
Could you please help add the certificate into tests/data_files.

Example PEM data:
-----BEGIN EC PRIVATE KEY-----
MIIBBgIBADAQBgcqhkjOPQIBBgUrgQQAIwSB3zCB3AIBAQRDAHxmkVH3bSQ6WHzE3Xz/kLPI6FfL
LDI1lMr0TlQ3YqCjHv4SN7ppi325fvM8f91S/po05YFSJggJ/nC2Z/hbNw4loAcGBSuBBAAjoYGJ
A4GGAAQAyYqoRNoHSYXPjAp5EoQubdbeZBB+1iZCn9gIvb8yoZFBUgl8JfPEqwKPRUhfOCrC+VaX
ICjrL1UkmM89Bhi2MfABachhEifGFExQ+InUe8VYlDoo/zV8C4/CTUF8x2l93tduVSXovL7r2fKw
0pmxIOU1nfsoJaz7swUX0A9x53K52cegDTALBgNVHQ8xBAMCAIA=
-----END EC PRIVATE KEY-----

gilles-peskine-arm commented:
The `data_files` directory is now in the framework submodule, so you need to make a companion pull request to https://github.com/Mbed-TLS/mbedtls-framework.

Please add both the commands to generate the new files to `data_files/Makefile`, and the new files themselves.

BhanuPrakash-P commented:
> The `data_files` directory is now in the framework submodule, so you need to make a companion pull request to https://github.com/Mbed-TLS/mbedtls-framework.
> 
> Please add both the commands to generate the new files to `data_files/Makefile`, and the new files themselves.

I am checking how to generate the ecdsa private key with attributes field using openssl. 
If openssl doesn't provide a way, then what would be the best option to proceed with the merge?

gilles-peskine-arm commented:
Please document how you generated the file. We like generating files with openssl because then they demonstrate interoperability. Python with `cryptography` would also be fine. But if that's not possible, you can use mbedtls tools or manual fiddling (e.g. with perl). We can potentially add other tools too, though it's more complicated.

What we want is:

1. To know what we have.
2. To be able to re-create a modified file, e.g. with a different expiration date, or testing a different edge case.

BhanuPrakash-P commented:
> Please document how you generated the file. We like generating files with openssl because then they demonstrate interoperability. Python with `cryptography` would also be fine. But if that's not possible, you can use mbedtls tools or manual fiddling (e.g. with perl). We can potentially add other tools too, though it's more complicated.
> 
> What we want is:
> 
> 1. To know what we have.
> 2. To be able to re-create a modified file, e.g. with a different expiration date, or testing a different edge case.

Please find the details below.
> 1. To know what we have.

The private key shared earlier was generated by a .NET server application. We have client applications which uses openssl, mbedtls. With openssl, certificate parsing is fine. But with mbedtls, we are facing the parsing issues(as mentioned in #9422).

> 2. To be able to re-create a modified file, e.g. with a different expiration date, or testing a different edge case.

We could not find openssl commands which generates the private key with 'Attributes' field. Is it okay to add the shared private key into data_files and add an entry in Makefile?

gilles-peskine-arm commented:
> The private key shared earlier was generated by a .NET server application.

Ok. Then I guess we're just going to have to take it as is. (It's not as bad for a key as for a certificate where we might want to change the dates and the signing CA.)

To make this clear to future maintainers, I think it would be best to commit the file into a new subdirectory `data_files/interoperability`.

Please also (or only) commit the key in DER format, so that we can run that parsing test even in configurations where PEM support is disabled. If you commit the PEM format as well, there should be a makefile rule to convert one into the other.

BhanuPrakash-P commented:
> > The private key shared earlier was generated by a .NET server application.
> 
> Ok. Then I guess we're just going to have to take it as is. (It's not as bad for a key as for a certificate where we might want to change the dates and the signing CA.)
> 
> To make this clear to future maintainers, I think it would be best to commit the file into a new subdirectory `data_files/interoperability`.
> 
> Please also (or only) commit the key in DER format, so that we can run that parsing test even in configurations where PEM support is disabled. If you commit the PEM format as well, there should be a makefile rule to convert one into the other.

test suite is updated with the private key.

gilles-peskine-arm commented:
Thanks! Can you please add a comment explaining how the key was obtained? Ideally some code for .NET, but at least explanations that a human could follow.

BhanuPrakash-P commented:
@gilles-peskine-arm, This PR is waiting on "TF OpenCI: PR tests Expected — Waiting for status to be reported".
I Could not find the reason or logs for the same. Could you please help in identifying the problem?

Thanks!

tom-cosgrove-arm commented:
> @gilles-peskine-arm, This PR is waiting on "TF OpenCI: PR tests Expected — Waiting for status to be reported". I Could not find the reason or logs for the same. Could you please help in identifying the problem?
> 
> Thanks!

It means that the CI run needs to be started on this PR. I've kicked it off

BhanuPrakash-P commented:
@gilles-peskine-arm, CI is completed for this PR. can you please remove [needs-ci](https://github.com/Mbed-TLS/mbedtls/labels/needs-ci) and add [needs-review](https://github.com/Mbed-TLS/mbedtls/labels/needs-review)  [needs-reviewer](https://github.com/Mbed-TLS/mbedtls/labels/needs-reviewer) labels.

Thanks!

BhanuPrakash-P commented:
this pull request is ready for review.

BhanuPrakash-P commented:
Hi @tom-cosgrove-arm, Do I need to create a separate backporting PR for this PR to get reviewed?

tom-cosgrove-arm commented:
@minosgalanakis Can I hand over to you to answer [the OP's question](https://github.com/Mbed-TLS/mbedtls/pull/9423#issuecomment-2468276186)

minosgalanakis commented:
@BhanuPrakash-Crestron thank you for your contribution. Having a backport is not a strong requirement for a pr to be reviewed, but rather the team's reviewe capacity.  But since the pr is a fixing an issue which is present in 3.6 it will need a backport before it can be merged.
